api/v1/api_test.go

```diff
@@ -0,0 +1,273 @@
+						Value: proto.String("basevalue2"),
+					},
+					{
+						Name:  proto.String("instance"),
+						Value: proto.String("instance2"),
+					},
+					{
+						Name:  proto.String("job"),
+						Value: proto.String("job1"),
+					},
+					{
+						Name:  proto.String("labelname"),
+						Value: proto.String("val2"),
+					},
+				},
+				Gauge: &dto.Gauge{
+					Value: proto.Float64(math.Inf(+1)),
+				},
+			},
+			{
+				Label: []*dto.LabelPair{
+					{
+						Name:  proto.String("instance"),
+						Value: proto.String("instance2"),
+					},
+					{
+						Name:  proto.String("job"),
+						Value: proto.String("job1"),
+					},
+					{
+						Name:  proto.String("labelname"),
+						Value: proto.String("val1"),
+					},
+				},
+				Gauge: &dto.Gauge{
+					Value: proto.Float64(math.Inf(-1)),
+				},
+			},
+		},
+	}
+
+	grouping1 = map[string]string{
+		"job":      "job1",
+		"instance": "instance1",
+	}
+)
+
+func templateMetrics(metricStore storage.MetricStore) []interface{} {
+	familyMaps := metricStore.GetMetricFamiliesMap()
+	res := make([]interface{}, 0)
+	for _, v := range familyMaps {
+		metricResponse := make(map[string]interface{})
+		metricResponse["labels"] = v.Labels
+		metricResponse["last_push_successful"] = v.LastPushSuccess()
+		for name, metricValues := range v.Metrics {
+			metricFamily := metricValues.GetMetricFamily()
+			uniqueMetrics := metrics{
+				Type:      metricFamily.GetType().String(),
+				Help:      metricFamily.GetHelp(),
+				Timestamp: metricValues.Timestamp,
+				Metrics:   makeEncodableMetrics(metricFamily.GetMetric(), metricFamily.GetType()),
+			}
+			metricResponse[name] = uniqueMetrics
+		}
+		res = append(res, metricResponse)
+	}
+	return res
+}
+
+func compareMaps(mainMap map[string]interface{}, testMap map[string]string) bool {
+	if len(mainMap) != len(testMap) {
+		return false
+	}
+
+	for key, _ := range mainMap {
+		if mainMap[key].(string) != testMap[key] {
+			return false
+		}
+	}
+	return true
+}
+
+func compareArray(a, b []byte) bool {
+	if len(a) != len(b) {
+		return false
+	}
+	for i, v := range a {
+		if v != b[i] {
+			return false
+		}
+	}
+	return true
+}
+
+func TestStatusAPI(t *testing.T) {
+	dms := storage.NewDiskMetricStore("", 100*time.Millisecond, nil, logger)
+	TestApi := New(logger, dms, testFlags, testBuildInfo)
+
+	req, err := http.NewRequest("GET", "http://example.org/", &bytes.Buffer{})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	w := httptest.NewRecorder()
+
+	testResponse := response{}
+	TestApi.status(w, req)
+	json.Unmarshal(w.Body.Bytes(), &testResponse)
+	jsonData := testResponse.Data.(map[string]interface{})
+	responseFlagData := jsonData["flags"].(map[string]interface{})
+	responseBuildInfo := jsonData["build_information"].(map[string]interface{})
+
+	if expected, got := http.StatusOK, w.Code; expected != got {
+		t.Errorf("Wanted status code %v, got %v.", expected, got)
+	}
+
+	if !compareMaps(responseFlagData, testFlags) {
```

Comment:
Would `reflect.DeepEqual` work here?

Reply:
Probably reflect.DeepEqual was not giving correct answer

Reply:
`reflect.DeepEqual` fails because the two maps have a different type.

Rather than `compareMaps`, how about defining a function like
```Go
func convertMap(m map[string]string) map[string]interface{} {
	result := map[string]interface{}{}
	for k, v := range m {
		result[k] = v
	}
	return result
}
```

This will then allow to write this line as
```Go
if !reflect.DeepEqual(responseFlagData, convertMap(testFlags)) {
```

It's in sum not much less code than your approach, but I suspect that many readers will ask themselves why you didn't use `reflect.DeepEqual` and suspect a tricky reason. This makes it clearer that it is really just a plain comparison.